REGRESSION(267849@main): MESSAGE_CHECK() failure when mouse events are being handled during a process swap

charliewolfe opened the pull request:
#### 0172e5a8bf88fbaed136e070c6709cba6dc93ff5
<pre>
REGRESSION(267849@main): MESSAGE_CHECK() failure when mouse events are being handled during a process swap
<a href="https://bugs.webkit.org/show_bug.cgi?id=263175">https://bugs.webkit.org/show_bug.cgi?id=263175</a>
rdar://116942710

Reviewed by Wenson Hsieh.

On a process swap, events queued in the UI process will be cleared. This was needed before 267849@main,
because the `didReceiveEvent()` IPC messages would not be sent to dequeue the events after terminating
the source web process. After 267849@main, it’s possible for the completion handler to be called after
clearing the event queues. So the message checks in `didReceiveEvent()` expecting the event queues to
not be empty will fail. We can expect the completion handler to be called for crashed web processes, so
we shouldn’t need to clear the event queues on process termination anymore.

We should also send the next queued mouse/touch/key events even after the web process crashes.

Added `ProcessSwap.MouseEventDuringCrossSiteProvisionalNavigation` to verify the web process no longer
crashes when a mouse event is being handled during a cross-site provisional navigation.

* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::sendMouseEvent):
(WebKit::WebPageProxy::sendKeyEvent):
(WebKit::WebPageProxy::touchEventHandlingCompleted):
(WebKit::WebPageProxy::handleTouchEvent):
(WebKit::WebPageProxy::mouseEventHandlingCompleted):
(WebKit::WebPageProxy::keyEventHandlingCompleted):
(WebKit::WebPageProxy::didReceiveEvent):
(WebKit::WebPageProxy::resetStateAfterProcessExited):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ProcessSwapOnNavigation.mm:

Canonical link: <a href="https://commits.webkit.org/269366@main">https://commits.webkit.org/269366@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2b44246e7325c4c11c3008799ffd42a9f10be1d4

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/22385 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/48/builds/3 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/23463 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/24289 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/20719 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/22639 "Passed tests") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/49/builds/3 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/22918 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/21730 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/22624 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/49/builds/3 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/19434 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/25144 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/49/builds/3 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/20289 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/26541 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/49/builds/3 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/20526 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/24395 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/4 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/17853 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/4 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/5 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/2796 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/5 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

whsieh commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -3596,25 +3596,21 @@ void WebPageProxy::sendKeyEvent(const NativeWebKeyboardEvent& event)
-            return;
-        if (!eventType)
+        if (!m_pageClient || !eventType) {
+            internals().keyEventQueue.removeFirst();
```
Should this still `MESSAGE_CHECK` if `internals().mouseEventQueue.isEmpty()`?

The implementation of `removeFirst()` looks like it would otherwise crash the UI process in the case where it's empty:

```
template<typename T, size_t inlineCapacity>
inline void Deque<T, inlineCapacity>::removeFirst()
{
    checkValidity();
    invalidateIterators();
    RELEASE_ASSERT(!isEmpty());
```

charliewolfe replied:
Yeah we should probably have the same `MESSAGE_CHECK` here as well. Nice catch, will fix.

whsieh commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -3896,10 +3892,10 @@ void WebPageProxy::handleTouchEvent(const NativeWebTouchEvent& event)
-                return;
-            if (!eventType)
+            if (!m_pageClient || !eventType) {
+                internals().touchEventQueue.removeFirst();
```
Ditto.

whsieh commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -8642,6 +8658,70 @@ void WebPageProxy::setCursorHiddenUntilMouseMoves(bool hiddenUntilMouseMoves)
+    }
+
+    if (!internals().mouseEventQueue.isEmpty()) {
+        LOG(MouseHandling, " UIProcess: handling a queued mouse event from didReceiveEvent");
```
Nit - does this log message need to be updated? (Since `mouseEventHandlingCompleted` can be called from outside of `didReceiveEvent`)

whsieh commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -8642,6 +8658,70 @@ void WebPageProxy::setCursorHiddenUntilMouseMoves(bool hiddenUntilMouseMoves)
+    bool canProcessMoreKeyEvents = !internals().keyEventQueue.isEmpty();
+    if (canProcessMoreKeyEvents && m_mainFrame) {
+        auto nextEvent = internals().keyEventQueue.first();
+        LOG(KeyHandling, " UI process: sent keyEvent from didReceiveEvent");
```
Ditto.

charliewolfe committed:
REGRESSION(267849@main): MESSAGE_CHECK() failure when mouse events are being handled during a process swap
https://bugs.webkit.org/show_bug.cgi?id=263175
rdar://116942710

Reviewed by Wenson Hsieh.

On a process swap, events queued in the UI process will be cleared. This was needed before 267849@main,
because the `didReceiveEvent()` IPC messages would not be sent to dequeue the events after terminating
the source web process. After 267849@main, it’s possible for the completion handler to be called after
clearing the event queues. So the message checks in `didReceiveEvent()` expecting the event queues to
not be empty will fail. We can expect the completion handler to be called for crashed web processes, so
we shouldn’t need to clear the event queues on process termination anymore.

We should also send the next queued mouse/touch/key events even after the web process crashes.

Added `ProcessSwap.MouseEventDuringCrossSiteProvisionalNavigation` to verify the web process no longer
crashes when a mouse event is being handled during a cross-site provisional navigation.

* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::sendMouseEvent):
(WebKit::WebPageProxy::sendKeyEvent):
(WebKit::WebPageProxy::touchEventHandlingCompleted):
(WebKit::WebPageProxy::handleTouchEvent):
(WebKit::WebPageProxy::mouseEventHandlingCompleted):
(WebKit::WebPageProxy::keyEventHandlingCompleted):
(WebKit::WebPageProxy::didReceiveEvent):
(WebKit::WebPageProxy::resetStateAfterProcessExited):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ProcessSwapOnNavigation.mm:

Canonical link: https://commits.webkit.org/269366@main
